### DISCUSSION TITLE
Is it possible to model storage tank in a gas network using pandapipes?

### DISCUSSION THREAD

**Discussion by akhiljoseph007** [e2nIEE/pandapipes#397] [General]
We have a 'create_storage' function in pandapower. I would like to model a similar storage facility in a gas network with a hydrogen storage tank (Interested to know about the change in Level of Gas) using pandapipes. Is there any way I can model a storage for a time series analysis? Thanks.

**Comment by jkisse** [comment]
Currently, no storage component is implemented yet. There is an [old pull request](https://github.com/e2nIEE/pandapipes/pull/233) for a basic storage component, but it still needs a bit of work. 
Do you know [the example in pandapower](https://github.com/e2nIEE/pandapower/blob/develop/tutorials/building_a_controller.ipynb) that models a battery storage by a controller? Something similar would be required in pandapipes as well.

**Comment by bpampel** [comment]
I'd also find such a feature desirable. I'm trying to model freshwater networks, and these typically come with storage facilities.
Pandapipes provides already a quite good approach to modelling these networks in combination with pandas and networkx, which I'd love to utilize for my projects. To use the models also with other tools, having some (simple) storage objects would be required.

Because I'd rather not patch in my own class, I'd also be happy to contribute to the existing PR which seems to be stale since some time: What is breaking the progress there and are there ways I could help?

**Reply by jkisse** [reply]
The component in PR #233 is quite simplistic, like a sink/source with a "state of charge" in kg. 
We are currently unsure, whether it might be too simple and more advanced calculations are required, like temperature calculation of the stored fluid or storage pressure. So the question is, whether a pure "mass storage" (+ example controller for the charging/discharging behavior) would be useful (the users could add their pressure calculation, if required), and "thermal storage" should be considered as a completely different component. 
What are your thoughts on this, @bpampel and @akhiljoseph007?

**Reply by bpampel** [reply]
At least from my side a pure mass storage unit with definable discharge behavior would probably be sufficient for what I plan to do.

The simplest discharge model is probably based on [Torricelli's law](https://en.wikipedia.org/wiki/Torricelli%27s_law), which means that the flow at the connecting node is proportional to the square root of the height of the stored water column (assuming a constant cross-section): `q = \sqrt{2 * g * h}`.
For simple applications you typically multiply this by a factor C < 1 ("discharge coefficient" or german "Ausflussbeiwert") that takes the viscosity of the fluid and the shape of the exit hole into account. Not an expert there, but I think this is mostly an empirical value, so this would be a parameter of the storage object to be set by the user.
To calculate the heigt of the column (or equally the pressure) you'd need to set the floor area as additional parameter of the tank.

All other parameters besides these two (discharge coefficient + area) would be optional for the simplest model, but depending on the use case you might want additional ones, e.g. maximum storage capacity.

But I also see your point about thermal storage. For our water applications we'll neglect thermal effects and just assume constant temperature everywhere and at all times, which makes these thoughts about the temperature of the stored fluid irrelevant. Of course for non-constant temperature flow one would need to take the "mixing" into account.
If you rather want to do it in a correct way directly instead of simple and "wrong", I can also understand that.

**Reply by jkisse** [reply]
Hi @bpampel and @akhiljoseph007,

the old PR for storage units has been updated. It is the simplest version of a storage, so you might have to add additional parameters like floor area, discharge coefficients etc. as additional arguments. There is also a tutorial on the implementation of a Storage Controller [here](https://github.com/jkisse/pandapipes/blob/add_storage/tutorials/building_a_storage_controller.ipynb).

We expect that the requirements for a storage component vary a lot among the different use cases and thus, the Controllers are not provided in the `pandapipes.control` module itself. Anyway, more advanced storage models can inherit from this mass_storage class and can also be added to pandapipes.

Please let me know if you can use the current implementation and if you suggest changes.

**Reply by bpampel** [reply]
Looks good to me on first sight, thanks!
I do not have the deep understanding of controllers yet, but I guess by using them it is possible to model the discharge rate (mdot_kg_s) as something that depends on the state of the storage unit (e.g. the pressure at the bottom of the tank) instead of it being a constant. Don't have the time to really dig into it currently, but I think then this would totally suit my purposes.
